### PR TITLE
Fix satint

### PR DESCRIPTION
**Pull request by kwxm**
There were some problems with the version of `SatInt` in `master`. For example:
```
> n = -1111111111111111111 :: SatInt
> n
-1111111111111111111  -- 19 1's: add another 1 on the end and you get minBound because it overflows
> 9*n
9223372036854775807 -- MaxBound, but should surely be MinBound
> m = 8*n
> m
-8888888888888888888 
> m+n
8446744073709551617  -- Should be MinBound


> a = 1111111111111111111 :: SatInt
> b = 8888888888888888888 :: SatInt
> a
1111111111111111111
> b
8888888888888888888
> 9*a
9223372036854775807   -- MaxBound
> a*b
9223372036854775807   -- MxBound
> a+b
-8446744073709551617  -- ???
> (-8)*a
-8888888888888888888 
> (-9)*a
9223372036854775807  -- ???
```

This contains contains some experiments to try and fix these.  It's quite delicate: small modifications can change the benchmark execution times noticeably.  The version here is about the fastest one I was able to come up with.  It's slightly faster than the version in master: see below.

I only looked at `plusSI`, `minusSI` and `timesSI`.  I'd be inclined to get rid of `quot`, `rem` and so on because we don't really need them,  they'd be very difficult to get right, and it's not clear what should happen when one or both of the arguments is `minBound`/`maxBound`.

This isn't supposed to be a proper PR, just some suggestions and comments.  It'd be good to have some tests when we have a final version.

Validation benchmark results:

```
                          master      this branch
---------------------------------------------------------
crowdfunding/1           1.628 ms	1.608 ms    -1.2%
crowdfunding/2           1.639 ms	1.629 ms    -0.6%
crowdfunding/3           1.636 ms	1.625 ms    -0.7%
crowdfunding/4           807.9 μs	803.5 μs    -0.5%
crowdfunding/5           807.5 μs	803.5 μs    -0.5%
future/1                 952.8 μs	945.4 μs    -0.8%
future/2                 1.919 ms	1.907 ms    -0.6%
future/3                 1.917 ms	1.907 ms    -0.5%
future/4                 2.376 ms	2.362 ms    -0.6%
future/5                 2.895 ms	2.883 ms    -0.4%
future/6                 2.431 ms	2.404 ms    -1.1%
future/7                 2.914 ms	2.878 ms    -1.2%
multisigSM/1             1.706 ms	1.678 ms    -1.6%
multisigSM/2             1.725 ms	1.700 ms    -1.4%
multisigSM/3             1.748 ms	1.722 ms    -1.5%
multisigSM/4             1.770 ms	1.758 ms    -0.7%
multisigSM/5             2.118 ms	2.104 ms    -0.7%
multisigSM/6             1.700 ms	1.689 ms    -0.6%
multisigSM/7             1.721 ms	1.712 ms    -0.5%
multisigSM/8             1.740 ms	1.730 ms    -0.6%
multisigSM/9             1.769 ms	1.756 ms    -0.7%
multisigSM/10            2.112 ms	2.102 ms    -0.5%
vesting/1                1.638 ms	1.628 ms    -0.6%
vesting/2                1.428 ms	1.420 ms    -0.6%
vesting/3                1.638 ms	1.626 ms    -0.7%
marlowe/trustfund/1      4.582 ms	4.540 ms    -0.9%
marlowe/trustfund/2      3.398 ms	3.347 ms    -1.5%
marlowe/zerocoupon/1     4.747 ms	4.710 ms    -0.8%
marlowe/zerocoupon/2     2.940 ms	2.920 ms    -0.7%
```

Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-core/plutus-core/src/Data/SatInt.hs [hunk 1]
```diff
@@ -23,6 +27,21 @@ fromSat (SI x) = x
 
 toSat :: Int -> SatInt
 toSat = SI
+{- **** ^ No range check:
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/Data/SatInt.hs:30]: Yes, we should not export these, and use `fromInteger` instead.

File: plutus-core/plutus-core/src/Data/SatInt.hs [hunk 2]
```diff
@@ -246,21 +265,52 @@ quotRemSI a@(I# _) b@(I# _) = (SI (a `quotInt` b), SI (a `remInt` b))
...
 
+
 plusSI :: SatInt -> SatInt -> SatInt
+plusSI (SI (I# x#)) (SI (I# y#)) =
+  case addIntC# x# y#  of
+    (# r#, f# #) ->
+        if isTrue# f# then  -- Overflow;  I think the value of f# is always 1 in this case
+                            -- and doesn't give us any other useful information
+        if      isTrue# ((x# ># 0#) `andI#` (y# ># 0#)) then maxBound
+        else if isTrue# ((x# <# 0#) `andI#` (y# <# 0#)) then minBound
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/Data/SatInt.hs:276]: I don't understand this: the documentation of `addIntC` says that `r#` is the sum truncated to fit in an Int... isn't that exactly what we want? I can see from your experiments that that's not working, but it seems very odd.

File: plutus-core/plutus-core/src/Data/SatInt.hs [hunk 3]
```diff
@@ -246,21 +265,52 @@ quotRemSI a@(I# _) b@(I# _) = (SI (a `quotInt` b), SI (a `remInt` b))
...
+        else 0  -- Shouldn't happen: error?  `succ maxBound` already raises an exception.
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/Data/SatInt.hs:277]: This is when x and y have opposite signs, right? I agree, I don't think that should be able to overflow.

**Review by michaelpj — COMMENTED** [review]
I'm still confused as to what's going on with `plusIntC` and friends, but I guess it's hard to tell... We should probably add some tests.

> I'd be inclined to get rid of quot, rem and so on because we don't really need them, they'd be very difficult to get right, and it's not clear what should happen when one or both of the arguments is minBound/maxBound.

Hm, maybe. There's also this discussion about what should happen with `minBound`: https://gitlab.haskell.org/ghc/ghc/-/issues/8695. But tbh, I don't think we care about this case, so we might as well just throw an error.

File: plutus-core/plutus-core/src/Data/SatInt.hs (hunk 2 above)
Comment (kwxm) [reply plutus-core/plutus-core/src/Data/SatInt.hs:276]: I now think we should be very careful about how we use `isTrue#`.  The documentation for `addInt#` says

> Add signed integers reporting overflow. First member of result is the sum truncated to an Int#; second member is zero if the true sum fits in an Int#, nonzero if overflow occurred (the sum is either too large or too small to fit in an Int#).

and for `isTrue#` it says

> Alias for tagToEnum#. Returns True if its parameter is 1# and False if it is 0#.

For the code above we've got `f#` and we want to branch if it's nonzero, but the documentation for `isTrue#` says nothing about what happens if its argument isn't `0#` or `1#`, and indeed the behaviour is weird:
```
> isTrue# 0#
False
> isTrue# 1#
True
> isTrue# 2#
False
> isTrue# 3#
True
> isTrue# 4#
True
> isTrue# 5#
False
> isTrue# 6#
True
> isTrue# 7#
True
> isTrue# 8#
True
> isTrue# 9#
True
> isTrue# 10#
True
> isTrue# 11#
False
> isTrue# 12#
False
> isTrue# 13#
False
> isTrue# 14#
False
> isTrue# 15#
False
> isTrue# 16#
False
```
Experiments with `addInt#` and `subInt#` suggests that the second component of the return value is always `0#` or `1#`, so this shouldn't matter; however all it says is that the second component is nonzero in case of overflow, and for all we know the exact value may depend on the CPU architecture or something.  Unfortunately `isTrue#` seems to be the only efficient way to check values;  there's not even an `isFalse#` (but see the note `Note [Optimizing isTrue#]` in [this file](https://hackage.haskell.org/package/ghc-prim-0.3.1.0/docs/src/GHC-Types.html)).

I think that to be on the safe side we really have to do something like 
```
if isTrue# (f# ==# 0#) then ...
```
to be safe (although I couldn't find an explicit statement of what `==#` is supposed to do.

Also we should probably say
```
if      isTrue# (x# ># 0#)  && isTrue# (y# ># 0#) then maxBound
```
instead of trying to be fancy with `and#`.  We'll only run this if there's an overflow, so it doesn't matter if it's not too efficient.

Comment (kwxm) [reply plutus-core/plutus-core/src/Data/SatInt.hs:276]: > I don't understand this: the documentation of `addIntC` says that `r#` is the sum truncated to fit in an Int... isn't that exactly what we want? I can see from your experiments that that's not working, but it seems very odd.

I think maybe it's truncated to 64 bits, not a (signed) `Int`.  The maximum value of `Int#` is 2^63 - 1 = 9223372036854775807,
and if we put a=1111111111111111111 then 9*a is bigger than this (and so overflows) but still fits into 64 bits, as a negative number.  

Up above, we had a+8*a = -8446744073709551617 (but 9*a wasn't equal to this!), and as integers 2^64-8446744073709551617 = 9999999999999999999, so I think there's some two's complement stuff going on.  I always get very confused by this stuff though.

The documentation for `addIntC#` says

> First member of result is the sum truncated to an Int#; 

and I think that must mean that what it's doing is equivalent to adding the numbers as integers and then forgetting everything except the bottom 64 bits (although I'm not sure how that explanation would work if negative numbers were involved).

Next it says

> second member is zero if the true sum fits in an Int#, nonzero if overflow occurred (the sum is either too large or too small to fit in an Int#).

I think it should really say "fit in a _signed_ Int'` there.

File: plutus-core/plutus-core/src/Data/SatInt.hs (hunk 3 above)
Comment (kwxm) [reply plutus-core/plutus-core/src/Data/SatInt.hs:277]: Yes, that was my thinking.

File: plutus-core/plutus-core/src/Data/SatInt.hs [hunk 4]
```diff
@@ -246,21 +265,52 @@ quotRemSI a@(I# _) b@(I# _) = (SI (a `quotInt` b), SI (a `remInt` b))
...
+ **** -}
 
 minusSI :: SatInt -> SatInt -> SatInt
 minusSI (SI (I# x#)) (SI (I# y#)) =
-  case subIntC# x# y# of
-    (# r#, _ #) -> SI (I# r#)
+    case subIntC# x# y# of
+      (# r#, f# #) ->
+          if isTrue# f# then -- Overflow
+          if      isTrue# ((x# >=# 0#) `andI#` (y# <# 0#)) then maxBound
```
[the remark this answers is not in the document]
Comment (kwxm) [reply plutus-core/plutus-core/src/Data/SatInt.hs:300]: > Any reason for the `>=` here versus `>` above?

I think that for `plusSi` we could equally well put `>=`, but that would only apply when one of the arguments is 0, and surely you can't get an overflow if you're adding 0 to something (although I'm beginning to have my doubts about that now).

We do need `>=` here because `0-minBound` should overflow (`minBound` is -2^63, `maxBound` is 2^63-1).  I think it'd be safe to change `<#`  into `<=#` here though.

Note that there's also a `negateInt#` operation, and the [documentation](https://hackage.haskell.org/package/ghc-prim-0.7.0/docs/GHC-Prim.html) for that says

> negateInt# of the most negative number is an identity operation. This way, negateInt# is always its own inverse.

In other words, `-minBound = minBound` (and both are negative), which I think also happens in Java, C, and so on.  This doesn't happen with the definitions in this file.  The `negate` instance gives `-minBound = maxBound`  and the code above gives `0-minBound = maxBound`.

**Comment by kwxm** [comment]
> [Re division operations] Hm, maybe. There's also this discussion about what should happen with `minBound`: https://gitlab.haskell.org/ghc/ghc/-/issues/8695. But tbh, I don't think we care about this case, so we might as well just throw an error.

I don't think we're going to need division-related operations for costing, and leaving them in might be a trap for the unwary user: addition and subtraction are already problematic enough.  The thing that worries me (and this even happens with subtraction) is that you could have a large arithmetic expression where some intermediate result is `minBound` or `maxBound` but then you carry out a division that turns it into something smaller and quite reasonable-looking without the user getting any warning.  For example you're in general not going to have `(a*b) `div` b` == b`, which is something you'd expect to be true.  It's like introducing infinity and then being able to divide it by 123 to get a finite number.

If we do leave these in then I think there should be a prominent warning in the file in case someone comes along in 5 years and thinks "Hh, this is just what I want!".

**Comment by michaelpj** [comment]
> The thing that worries me (and this even happens with subtraction) is that you could have a large arithmetic expression where some intermediate result is minBound or maxBound but then you carry out a division that turns it into something smaller and quite reasonable-looking without the user getting any warning.

You could say the same thing about subtraction! I agree that integers with saturating arithmetic are a little tricky, but I think that just means we should be careful with them, not that we should just not implement the operations (which do exactly what saturating operations are meant to!).

File: plutus-core/plutus-core/src/Data/SatInt.hs (hunk 2 above)
Comment (michaelpj) [reply plutus-core/plutus-core/src/Data/SatInt.hs:276]: Yikes. Okay, then I agree we need to do something like you suggest.

Comment (michaelpj) [reply plutus-core/plutus-core/src/Data/SatInt.hs:276]: > Experiments with addInt# and subInt# suggests that the second component of the return value is always 0# or 1#, so this shouldn't matter; however all it says is that the second component is nonzero in case of overflow, and for all we know the exact value may depend on the CPU architecture or something.

GHC itself has [tests](https://gitlab.haskell.org/ghc/ghc/-/blob/master/testsuite/tests/primops/should_run/T9430.hs#L65) for `addIntC` that seem to assume that it's indeed 0 or 1, so I think we can assume that. I'll ask in #ghc anyway.

**Comment by kwxm** [comment]
> You could say the same thing about subtraction!

I just did!  We don't really need subtraction either.  In restricting mode it seemed useful to count down from the initial budget and fail if cpu or memory drops below zero;  we could equally well count up from zero and fail if we reach the limit.  I think I argued in favour of counting down because I thought it might be faster to compare against zero rather than some big number.  That was when we were using Integers though.

**Comment by michaelpj** [comment]
I guess I also like having a full number type exposed. I was actually thinking of upstreaming `SatInt` into the `safe-int` package... but you're right that we don't really need it. We could effectively just give `SatInt` instances for `Monoid` and `Ord` and that would do the job.

I think I also argued for using subtraction, it seemed neater but I agree we don't need it.

So I see two routes: 
1. Leave all the operations in here, and just don't use them, relying on (say) a `Monoid` and `Ord` instance for `ExMemory` and `ExCPU`. 
2. Delete the questionable operations, leaving it to me to put them back if I ever get around to upstreaming.

**Comment by kwxm** [comment]
> So I see two routes:

It was your idea, so I'll leave it up to you.  I think it'd be OK to leave the other stuff in with a reminder of the issues.


On a slightly different note, all we really need to do is perform two additions (or maybe subtractions) and two comparisons per CEK step.  I wonder if it would be more efficient to expose functions which do these directly instead of via typeclasses. That's probably not a good idea because we have the `#if` that switches to `Integer` in the 32-bit case, but it might be worth checking.

**Comment by michaelpj** [comment]
Yeah, although I think this stuff should all inline pretty well. Worth checking, though.

**Comment by michaelpj** [comment]
I checked: currently `restricting` gets everything inlined right the way down to occurences of `subIntC` and `GHC.Prim.<#`, so that's all good.

**Comment by michaelpj** [comment]
I'm opening a version of this with some tests in a bit.
